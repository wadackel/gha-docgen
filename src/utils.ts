import * as fs from 'node:fs/promises';
import type { ZodError } from 'zod';
import { fromZodError } from 'zod-validation-error';

/**
 * Filesystem
 */
export const read = async (filepath: string): Promise<[Error, null] | [null, string]> => {
  try {
    const data = await fs.readFile(filepath, 'utf8');
    return [null, data];
  } catch (e) {
    return [e instanceof Error ? e : new Error(String(e)), null];
  }
};

export const write = async (filepath: string, data: string): Promise<Error | null> => {
  try {
    await fs.writeFile(filepath, data, 'utf8');
    return null;
  } catch (e) {
    return e instanceof Error ? e : new Error(String(e));
  }
};

/**
 * Error handling
 */
export const generateZodErrorMessage = (error: ZodError): string => {
  return fromZodError(error, {
    prefix: undefined,
    issueSeparator: ' | ',
    unionSeparator: ' OR ',
  }).toString();
};

/**
 * Strings
 */
const NEWLINE_REGEX = /\r\n|\r|\n/g;

export const nl2space = (input: string): string => {
  return input.replace(NEWLINE_REGEX, ' ');
};

export const nl2md = (input: string): string => {
  return input.replace(NEWLINE_REGEX, '  \n');
};

export const nl2br = (input: string): string => {
  return input.replace(NEWLINE_REGEX, '<br />');
};
