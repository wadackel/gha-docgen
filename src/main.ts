import * as path from 'node:path';
import type { IDebugger } from 'debug';
import YAML from 'yaml';
import type { DocgenStyle } from './schema';
import { actionSchema } from './schema';
import { generateZodErrorMessage, read, write } from './utils';
import { docgen } from './docgen';

export type MainOptions = {
  debug: IDebugger;
  cwd: string;
  files: string[];
  action: string | undefined;
  style: DocgenStyle;
};

export const main = async ({ debug, cwd, ...opts }: MainOptions): Promise<void> => {
  // Read action file
  let actionStr: string;

  if (opts.action !== undefined) {
    const filepath = path.join(cwd, opts.action);
    debug('read specified action file: "%s"', filepath);
    const [err, data] = await read(filepath);
    if (err !== null) {
      throw new Error(`Error reading action file: ${err.message}`);
    }
    actionStr = data;
  } else {
    // 1. read 'action.yml'
    let filepath = path.join(cwd, 'action.yml');
    debug('read default action file: "%s"', filepath);
    const [err1, data1] = await read(filepath);
    if (data1 !== null) {
      actionStr = data1;
    } else {
      // 2. read 'action.yaml'
      debug(err1);
      filepath = path.join(cwd, 'action.yaml');
      debug('read default action file: "%s"', filepath);
      const [err2, data2] = await read(filepath);
      if (err2 !== null) {
        throw new Error(`Error reading action file: ${err2.message}`);
      }
      actionStr = data2;
    }
  }

  debug('found action file: %s', actionStr);

  // Parse action file
  let actionObj: unknown;
  try {
    actionObj = YAML.parse(actionStr);
  } catch (e) {
    debug(e);
    const msg = e instanceof Error ? e.message : String(e);
    throw new Error(`Error parsing action file: ${msg}`);
  }

  const action = actionSchema.safeParse(actionObj);
  if (!action.success) {
    debug('action schema error: %s', action.error);
    throw new Error(`Error parsing action file: ${generateZodErrorMessage(action.error)}`);
  }

  // Read markdown files
  for (const filename of opts.files) {
    const filepath = path.join(cwd, filename);
    debug('read markdown file: "%s"', filepath);
    const [readErr, before] = await read(filepath);
    if (readErr !== null) {
      debug(readErr);
      throw new Error(`Error reading markdown file: ${readErr.message}`);
    }

    debug('generate docs for "%s"', filepath);
    debug('before: "%s"', before);
    const after = docgen(before, action.data, {
      style: opts.style,
    });
    debug('after: "%s"', after);

    const writeErr = await write(filepath, after);
    if (writeErr !== null) {
      debug(writeErr);
      throw new Error(`Error writing markdown file: ${writeErr.message}`);
    }
  }
};
