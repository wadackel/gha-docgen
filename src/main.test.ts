import * as path from 'node:path';
import { fileURLToPath } from 'node:url';
import { expect, test, vi } from 'vitest';
import { read, write } from './utils';
import { main } from './main';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

const FIXTURES_DIR = path.resolve(__dirname, '../fixtures');

test.each<[string, string, string[], string | undefined]>(
  // prettier-ignore
  [
    [
      'action.yml',
      path.join(FIXTURES_DIR, 'action-yml'),
      ['README.md'],
      undefined,
    ],
    [
      'action.yaml',
      path.join(FIXTURES_DIR, 'action-yaml'),
      ['README.md'],
      undefined,
    ],
    [
      'different paths',
      path.join(FIXTURES_DIR, 'action-yml-path'),
      ['README.md'],
      'dir/action.yml',
    ],
    [
      'different paths',
      path.join(FIXTURES_DIR, 'action-yml-path'),
      ['README.md'],
      'dir/action.yml',
    ],
    [
      'multi files',
      path.join(FIXTURES_DIR, 'multi-files'),
      ['md1.md', 'md2.md'],
      undefined,
    ],
  ],
)('%s', async (_, cwd, files, action) => {
  await Promise.all(
    files.map(async (file) => {
      const data = `
# Title

<!-- gha-description-start -->
<!-- gha-description-end -->

# Inputs

<!-- gha-inputs-start -->
<!-- gha-inputs-end -->

# Outputs

<!-- gha-outputs-start -->
<!-- gha-outputs-end -->
    `.trim();
      const error = await write(path.join(cwd, file), data);
      if (error !== null) {
        throw error;
      }
    }),
  );

  await main({
    debug: vi.fn() as never,
    cwd,
    files,
    action,
    style: 'section:h3',
  });

  const results = await Promise.all(
    files.map(async (file) => {
      const [error, data] = await read(path.join(cwd, file));
      if (error !== null) {
        throw error;
      }
      return data;
    }),
  );

  expect(results.join('\n')).toMatchSnapshot();
});

test('default action.yml not exists', async () => {
  await expect(
    main({
      debug: vi.fn() as never,
      cwd: FIXTURES_DIR,
      files: ['README.md'],
      action: undefined,
      style: 'section:h3',
    }),
  ).rejects.toThrowError(/Error reading action file:/);
});

test('specified action.yml not exists', async () => {
  await expect(
    main({
      debug: vi.fn() as never,
      cwd: FIXTURES_DIR,
      files: ['README.md'],
      action: 'not-found-file.yml',
      style: 'section:h3',
    }),
  ).rejects.toThrowError(/Error reading action file:/);
});

test('invalid action.yml schema', async () => {
  await expect(
    main({
      debug: vi.fn() as never,
      cwd: path.join(FIXTURES_DIR, 'invalid-action-yml'),
      files: ['README.md'],
      action: undefined,
      style: 'section:h3',
    }),
  ).rejects.toThrowError(
    'Error parsing action file: Invalid input: expected string, received undefined at "description" | Invalid input: expected object, received string at "inputs.github_token"',
  );
});

test('files not exists', async () => {
  await expect(
    main({
      debug: vi.fn() as never,
      cwd: path.join(FIXTURES_DIR, 'invalid-files'),
      files: ['README.md'],
      action: undefined,
      style: 'section:h3',
    }),
  ).rejects.toThrowError(/Error reading markdown file:/);
});
