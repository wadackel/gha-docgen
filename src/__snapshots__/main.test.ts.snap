// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`action.yaml 1`] = `
"# Title

<!-- gha-description-start -->
line1  
line2
<!-- gha-description-end -->

# Inputs

<!-- gha-inputs-start -->
### \`github_token\`

**Required:** \`false\`  
**Default:** \`\${{ github.token }}\`

The GitHub authentication token
<!-- gha-inputs-end -->

# Outputs

<!-- gha-outputs-start -->

<!-- gha-outputs-end -->"
`;

exports[`action.yml 1`] = `
"# Title

<!-- gha-description-start -->
line1  
line2
<!-- gha-description-end -->

# Inputs

<!-- gha-inputs-start -->
### \`github_token\`

**Required:** \`false\`  
**Default:** \`\${{ github.token }}\`

The GitHub authentication token
<!-- gha-inputs-end -->

# Outputs

<!-- gha-outputs-start -->

<!-- gha-outputs-end -->"
`;

exports[`different paths 1`] = `
"# Title

<!-- gha-description-start -->
line1  
line2
<!-- gha-description-end -->

# Inputs

<!-- gha-inputs-start -->
### \`github_token\`

**Required:** \`false\`  
**Default:** \`\${{ github.token }}\`

The GitHub authentication token
<!-- gha-inputs-end -->

# Outputs

<!-- gha-outputs-start -->

<!-- gha-outputs-end -->"
`;

exports[`different paths 2`] = `
"# Title

<!-- gha-description-start -->
line1  
line2
<!-- gha-description-end -->

# Inputs

<!-- gha-inputs-start -->
### \`github_token\`

**Required:** \`false\`  
**Default:** \`\${{ github.token }}\`

The GitHub authentication token
<!-- gha-inputs-end -->

# Outputs

<!-- gha-outputs-start -->

<!-- gha-outputs-end -->"
`;

exports[`multi files 1`] = `
"# Title

<!-- gha-description-start -->
line1  
line2
<!-- gha-description-end -->

# Inputs

<!-- gha-inputs-start -->
### \`github_token\`

**Required:** \`false\`  
**Default:** \`\${{ github.token }}\`

The GitHub authentication token
<!-- gha-inputs-end -->

# Outputs

<!-- gha-outputs-start -->

<!-- gha-outputs-end -->
# Title

<!-- gha-description-start -->
line1  
line2
<!-- gha-description-end -->

# Inputs

<!-- gha-inputs-start -->
### \`github_token\`

**Required:** \`false\`  
**Default:** \`\${{ github.token }}\`

The GitHub authentication token
<!-- gha-inputs-end -->

# Outputs

<!-- gha-outputs-start -->

<!-- gha-outputs-end -->"
`;
